<html><head><meta http-equiv="Content-Type" content="text/html; charset=ISO-8859-1"><title>Alignment</title><meta name="generator" content="DocBook XSL-NS Stylesheets V1.75.2"><link rel="home" href="index.html" title="Java TreeView User's Manual"><link rel="up" href="ch02.html" title="Chapter 2. Feature Reference"><link rel="prev" href="ch02s09.html" title="Karyoscope"><link rel="next" href="ch02s11.html" title="File Formats"></head><body bgcolor="white" text="black" link="#0000FF" vlink="#840084" alink="#0000FF"><div class="navheader"><table width="100%" summary="Navigation header"><tr><th colspan="3" align="center">Alignment</th></tr><tr><td width="20%" align="left"><a accesskey="p" href="ch02s09.html">Prev</a> </td><th width="60%" align="center">Chapter 2. Feature Reference</th><td width="20%" align="right"> <a accesskey="n" href="ch02s11.html">Next</a></td></tr></table><hr></div><div class="sect1" title="Alignment"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a name="ch2_alignment"></a>Alignment</h2></div></div></div>

<div class="figure"><a name="AlignmentFigure"></a><p class="title"><b>Figure 2.14. Alignment Component Layout</b></p><div class="figure-contents">

<div><img src="figures/alignment.gif" alt="Alignment Component Layout"></div>
</div></div><br class="figure-break">

<p>
The Alignment view is very good for browing large alignments, even if you don't have any expression data for them. There is a utility, "aln2cdt.pl" available in the helper-scripts package from website to facilitate the viewing of clustalw alignments.
</p>
<p>
The alignment view relies up the existence of a column named "ALN" containing all the sequence data, with IUPAC symbols for the amino acids, and non-symbol spacer characters such as dashes designating the gaps. The alignment view will render the matching sequence into a dendrogram-view like two-level display, complete with gene tree.
</p>

<p>
There are two PERL scripts in the helper-scripts package available from the website which should aid in the usage of alignment view, aln2cdt.pl which will create a cdt with the appropriate columns, and potentially a gtr fie if there is a dnd file available, and appendPCL.pl, which will allow you to append expression data in the pcl format to the cdt alignment file. 
</p>

<p>
For a detailed howto to assist in making your own alignments, please see the fgf receptor example on the website (http://jtreeview.sourceforge.net)
</p>

</div><div class="navfooter"><hr><table width="100%" summary="Navigation footer"><tr><td width="40%" align="left"><a accesskey="p" href="ch02s09.html">Prev</a> </td><td width="20%" align="center"><a accesskey="u" href="ch02.html">Up</a></td><td width="40%" align="right"> <a accesskey="n" href="ch02s11.html">Next</a></td></tr><tr><td width="40%" align="left" valign="top">Karyoscope </td><td width="20%" align="center"><a accesskey="h" href="index.html">Home</a></td><td width="40%" align="right" valign="top"> File Formats</td></tr></table></div></body></html>
