<html><head><meta http-equiv="Content-Type" content="text/html; charset=ISO-8859-1"><title>File Formats</title><meta name="generator" content="DocBook XSL-NS Stylesheets V1.75.2"><link rel="home" href="index.html" title="Java TreeView User's Manual"><link rel="up" href="ch02.html" title="Chapter 2. Feature Reference"><link rel="prev" href="ch02s10.html" title="Alignment"><link rel="next" href="ch03.html" title="Chapter 3. Troubleshooting"></head><body bgcolor="white" text="black" link="#0000FF" vlink="#840084" alink="#0000FF"><div class="navheader"><table width="100%" summary="Navigation header"><tr><th colspan="3" align="center">File Formats</th></tr><tr><td width="20%" align="left"><a accesskey="p" href="ch02s10.html">Prev</a> </td><th width="60%" align="center">Chapter 2. Feature Reference</th><td width="20%" align="right"> <a accesskey="n" href="ch03.html">Next</a></td></tr></table><hr></div><div class="sect1" title="File Formats"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a name="ch2_file"></a>File Formats</h2></div></div></div>



<div class="figure"><a name="FilesFigure"></a><p class="title"><b>Figure 2.15. File Formats in Java TreeView</b></p><div class="figure-contents"> 

<div><img src="figures/Files.gif" alt="File Formats in Java TreeView"></div>
</div></div><br class="figure-break">

<p>
Java TreeView uses two file formats to represent data, both of which are tab-delimited text. The third file format is an xml formatted file which holds settings information for persistence (see <a class="xref" href="ch02s02.html" title="Persistence">the section called &#8220;Persistence&#8221;</a>).
</p>

<div class="itemizedlist" title="Java TreeView File Formats"><p class="title"><b>Java TreeView File Formats</b></p><ul class="itemizedlist" type="disc"><li class="listitem"> <p> Generalized CDT File (.cdt and .pcl)</p> </li><li class="listitem"> <p> Tree File (.atr and .gtr)</p> </li><li class="listitem"> <p> XML Settings File (.jtv and global settings file, see <a class="xref" href="ch02s02.html#ch2_persistence_global" title="Global Settings">the section called &#8220;Global Settings&#8221;</a>) </p> </li></ul></div>

<p>
The use of tab-delimitted text makes these files easy to edit in spreadsheet programs such as Excel, as well as manipulate with other programs.
</p>
<div class="sect2" title="Minimal File Requirements"><div class="titlepage"><div><div><h3 class="title"><a name="ch2_file_min"></a>Minimal File Requirements</h3></div></div></div>

<p>
In order to view data in Java Treeview, there must be a Generalized CDT file. All other files are optional. The minimal GCDT file has a header row which contains the name of the unique id column, the name of the annotation column, and the names of the experiments, followed by one or more rows of per-gene data. Such a file can be created in Excel, and then saved as tab-delimited text.
</p>
</div>
<div class="sect2" title="Generalized CDT File"><div class="titlepage"><div><div><h3 class="title"><a name="ch2_file_cdt"></a>Generalized CDT File</h3></div></div></div>


<div class="figure"><a name="gcdtFigure"></a><p class="title"><b>Figure 2.16. Screenshot of Generalized CDT file in Excel</b></p><div class="figure-contents"> 

<div><img src="figures/gcdt.gif" alt="Screenshot of Generalized CDT file in Excel"></div>
</div></div><br class="figure-break">
<p>
The generalized CDT file is a straightforward generalization of the CDT and PCL file formats. In addition to expression data, this file can contain additional per-gene and per-array annotation in columns before the GWEIGHT column or in rows before the EWEIGHT row. For backwards compatibility, if the GWEIGHT column is missing Java TreeView assumes the data starts on the third column, or the fourth column if the first column has the header GID. Similarly, if the EWEIGHT row is missing Java TreeView assumes the data starts on the second row. As a general practice, it is a good idea to include the GWEIGHT column and EWEIGHT row. </p>
<p>
In addition, Java TreeView does special things with the first two or three columns. If the first column is GID, the second and third are assumed to be the unique ID and NAME columns. If the first column is anything other than GID, the first and second columns are assume to be the unique ID and NAME columns. The unique ID is used for gene list export, and for some matching purposes when necessary. The NAME column is displayed as per-gene annotation in the dendrogram and other views.
</p>
<p>
There are annotation column names with special meaning to Java TreeView, and are used for coordinates or to set the color of gene names. These special columns are described after the basic file format, and should be avoided as annotation names unless you want that specific behavior.
</p>
<div class="sect3" title="Formal Description of Generalized CDT File"><div class="titlepage"><div><div><h4 class="title"><a name="ch2_file_cdt_all"></a>Formal Description of Generalized CDT File</h4></div></div></div> 

<p>
A generalized CDT file is a tab-delimitted text file with the following specifications. The leftmost column and topmost row are reserved for headers. The file must contain at least two columns followed by a column with the header GWEIGHT, and at least one row followed by a row with the header EWEIGHT. Any rows and  columns before the EWEIGHT and GWEIGHT are treated as annotation, and any after are treated as data. If a data value is missing or cannot be converted into a number, it is treated as not found. The annotation is kept in string form, and parsed by views as appropriate.
</p>
</div>
<div class="sect3" title="Headers With Special Meaning"><div class="titlepage"><div><div><h4 class="title"><a name="ch2_file_cdt_colheaders"></a> Headers With Special Meaning </h4></div></div></div>

<p> Some headers have special meaning to particular displays. This is a catalog of headers and their meaning to particular displays.
</p>
<div class="table"><a name="id1582462"></a><p class="title"><b>Table 2.1. CDT Column Headers With Special Meaning </b></p><div class="table-contents">
<table summary="CDT Column Headers With Special Meaning " border="1"><colgroup><col><col><col></colgroup><thead><tr><th align="left"> Header</th><th align="left"> Display </th><th align="left"> Meaning </th></tr></thead><tbody><tr><td align="left"> FGCOLOR </td><td align="left"> Dendrogram </td><td align="left"> Color in which to render text for particular gene</td></tr><tr><td align="left"> BGCOLOR </td><td align="left"> Dendrogram </td><td align="left"> Color in which to render background  for particular gene </td></tr><tr><td align="left"> LEAF</td><td align="left"> Dendrogram</td><td align="left"> The TIME at which this branch should be terminated. Used to indicate apopotosis in cell lineages, as well as phyogenetic distance in sequence alignments. See also TIME header for tree files.</td></tr><tr><td align="left">CHROMOSOME</td><td align="left">Karyoscope</td><td align="left">The chromosome on which the gene is located, a natural number</td></tr><tr><td align="left">ARM</td><td align="left">Karyoscope</td><td align="left">The arm of the chromosome, either "L", "R", or "1" meaning left, "2" meaning right.</td></tr><tr><td align="left">POSITION</td><td align="left">Karyoscope</td><td align="left">The distance of the spot from the centromere in arbitrary units</td></tr><tr><td align="left"> GROUP</td><td align="left"> Dendrogram</td><td align="left"> Defines a partitioning of genes. The current (1.0.13) implementation of Dendrogram will insert a gap every time the GROUP value changes. 
At some point in the future, I may make it so that all genes with identical GROUP values are put in one cluster. </td></tr></tbody></table>
</div></div><br class="table-break">
<div class="table"><a name="id1582622"></a><p class="title"><b>Table 2.2. CDT Row Headers With Special Meaning </b></p><div class="table-contents">
<table summary="CDT Row Headers With Special Meaning " border="1"><colgroup><col><col><col></colgroup><thead><tr><th align="left"> Header</th><th align="left"> Display </th><th align="left"> Meaning </th></tr></thead><tbody><tr><td align="left"> FGCOLOR </td><td align="left"> Dendrogram </td><td align="left"> Color in which to render text of array name</td></tr><tr><td align="left"> BGCOLOR </td><td align="left"> Dendrogram </td><td align="left"> Color in which to render background of array name </td></tr><tr><td align="left"> GROUP</td><td align="left"> Dendrogram</td><td align="left"> Defines a partitioning of arrays. 
The current (1.0.13) implementation of Dendrogram will insert a gap every time the GROUP value changes. 
At some point in the future, I may make it so that all rows with identical GROUP values are put in one cluster. </td></tr></tbody></table>
</div></div><br class="table-break">


</div>
</div>

<div class="sect2" title="Coordinates Files for Karyoscope"><div class="titlepage"><div><div><h3 class="title"><a name="ch2_file_karyoscope"></a> Coordinates Files for Karyoscope </h3></div></div></div>

<p>
In order for Karyoscope to correctly display gene expression data by chromosome location, it needs to know where exactly to position each unique ID. To this end, it looks for annotation columns with the names "CHROMOSOME",  "ARM" and "POSITION", which designate the chromosome, arm and position of a particular gene. "CHROMOSOME" should be a natural number indicating which chromosome the unique ID is on, "ARM" should be either "R" or "L" indicating the arm, and "POSITION" should be a real number indicating how far from the centromere the unique ID is. There is really no restriction on the units for position; bp or kb are natural choices.
</p>
<p>
A coordinates file is simply a generalized CDT file which has such columns. The coordinates files supplied with Java TreeView do not contain any expression data; they consist entirely of the unique id column, the chromosome, arm and position columns, and the required GWEIGHT column. However, any other generalized CDT file with the correct columns can serve as a coordinates file.
</p>
</div>


<div class="sect2" title="Tree Files"><div class="titlepage"><div><div><h3 class="title"><a name="ch2_file_gtr_atr"></a>Tree Files</h3></div></div></div>

<div class="figure"><a name="gtrFigure"></a><p class="title"><b>Figure 2.17. Screenshot of a Tree File in Excel</b></p><div class="figure-contents"> 

<div><img src="figures/gtr.gif" alt="Screenshot of a Tree File in Excel"></div>
</div></div><br class="figure-break">
<p>
Traditionally, tree files have no header, and consist of four columns. Each row represents a node in either a gene tree, for the GTR file, or an array tree in the ATR file. For each row, the first column is the identifier of the node, the second column is the left child of the node, the third column is the right child, and the fourth column is the correlation between the left and right child. This fourth column is used by Java TreeView to determine the height of the node when rendering a tree.
</p>
<p>
By analogy to the CDT file, the tree files have been generalized in Java Treeview. Generalized tree files have a header line identifying the different columns. All generalized GTR/ATR files must have NODEID as the name of the first column. Tree files with any other string in the first row of the first column will be treated as legacy tree files. All of the rows will be treated as defining nodes, and the headers will be assigned the headers "NODEID", "LEFT", "RIGHT" and "CORRELATION". The meaning of these headers, and others, are described in <a class="xref" href="ch02s11.html#ch2_file_tr_colheaders" title="Tree File Headers">the section called &#8220; Tree File Headers &#8221;</a>.
</p>
<div class="sect3" title="Tree File Headers"><div class="titlepage"><div><div><h4 class="title"><a name="ch2_file_tr_colheaders"></a> Tree File Headers </h4></div></div></div>

<p> All tree file headers can be displayed as node annotation in the dendrogram view. After loading a file in java treeview, select "Settings-&gt;Annotations..." to open the annotations dialog. Click "Gene Nodes" or "Array Nodes" and then select the desired headers. If you then mouse over the tree, the annotation for the select node will appear in the info pane.
</p>
<p>
Headers that have meaning beyond being used as annotation are described here.
</p>
<div class="table"><a name="id1582791"></a><p class="title"><b>Table 2.3. GTR/ATR Column Headers With Special Meaning </b></p><div class="table-contents">
<table summary="GTR/ATR Column Headers With Special Meaning " border="1"><colgroup><col><col><col></colgroup><thead><tr><th align="left"> Header</th><th align="left"> Display </th><th align="left"> Meaning </th></tr></thead><tbody><tr><td align="left"> NODEID</td><td align="left"> Dendrogram</td><td align="left"> The value in this column serves as the identifier for the node. It must be unique.</td></tr><tr><td align="left"> LEFT</td><td align="left"> Dendrogram</td><td align="left"> The NODEID of the left child of this node. If the left child is not an internal node but a gene from the CDT file, the value should be the gene identifer, i.e. the value in the first column of the CDT file.</td></tr><tr><td align="left"> RIGHT</td><td align="left"> Dendrogram</td><td align="left"> The NODEID of the left child of this node, with leaf nodes handled as for the LEFT column.</td></tr><tr><td align="left"> CORRELATION</td><td align="left"> Dendrogram</td><td align="left"> The correlation value for this node. If this column appears, the implication is that the nodes in the tree should be arranged with a value of -1 most distal from the data matrix, with a value of 1 immediately proximal. This column and the TIME column should not appear in the same tree file.</td></tr><tr><td align="left"> TIME</td><td align="left"> Dendrogram</td><td align="left"> <p>The time at which this node branched. Java Treeview is used by some  to display cell lineages. For this purpose, it is useful to arrange branches in the tree by time at which the cell divided. This column is also used for indicating branching in sequence alignments, although TIME means phyogenetic divergence time in that context.</p><p> If this column appears, the implication is that the nodes in the tree should be arrange with a value of 0 most distal from the data matrix, and that the node with maximum time value should be immediately adjacent to the data. </p></td></tr><tr><td align="left"> NODECOLOR</td><td align="left"> Dendrogram</td><td align="left"> <p>The color value for this node. In the abscence of this column, the dendrogram is rendered black when not selected. Even when this column appears, selected nodes are rendered red.</p></td></tr></tbody></table>
</div></div><br class="table-break">
</div>
</div>
</div><div class="navfooter"><hr><table width="100%" summary="Navigation footer"><tr><td width="40%" align="left"><a accesskey="p" href="ch02s10.html">Prev</a> </td><td width="20%" align="center"><a accesskey="u" href="ch02.html">Up</a></td><td width="40%" align="right"> <a accesskey="n" href="ch03.html">Next</a></td></tr><tr><td width="40%" align="left" valign="top">Alignment </td><td width="20%" align="center"><a accesskey="h" href="index.html">Home</a></td><td width="40%" align="right" valign="top"> Chapter 3. Troubleshooting</td></tr></table></div></body></html>
